esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -1241,194 +1241,200 @@ void setDeleteFileResults(const char* fileName, const char* nodeGroup, bool fail
-}
+                const char *lfn = dfsLFN.get();
+                Owned<IDistributedFile> df = fdir.lookup(lfn, userdesc, true, false, false, nullptr, defaultPrivilegedUser, 30000); // 30 sec timeout
+                if (!df)
+                {
+                    VStringBuffer message("Could not delete file %s: file not found.", lfn);
+                    addResult(lfn, group, true, message);
+                    continue;
+                }
 
-bool CWsDfuEx::DFUDeleteFiles(IEspContext &context, IEspDFUArrayActionRequest &req, IEspDFUArrayActionResponse &resp)
-{
-    if (isDetachedFromDali())
-        throw MakeStringException(ECLWATCH_INVALID_INPUT, "ESP server is detached from Dali. Please try later.");
+                Owned<IDistributedSuperFileIterator> superOwners = df->getOwningSuperFiles();
+                if (!removeFromSuperFiles && superOwners->first())
+                {
+                    VStringBuffer message("The file %s can not be deleted because it is owned by one or more superfiles.", lfn);
+                    addResult(lfn, group, true, message);
+                    continue;
+                }
 
-    double version = context.getClientVersion();
-    Owned<IUserDescriptor> userdesc;
-    const char *username = context.queryUserId();
-    if(username && *username)
-    {
-        userdesc.setown(createUserDescriptor());
-        userdesc->set(username, context.queryPassword(), context.querySignature());
-    }
+                bool fileCanBeDeleted = true;
+                ForEach(*superOwners)
+                {
+                    IDistributedSuperFile &superOwner = superOwners->query();
+                    try
+                    {
+                        superOwner.removeSubFile(lfn, false, false, nullptr);
+                        PROGLOG("File %s is removed from superfile %s", lfn, superOwner.queryLogicalName());
+                    }
+                    catch(IException *e)
+                    {
+                        VStringBuffer message("Could not delete file %s from superfile %s: ", lfn, superOwner.queryLogicalName());
+                        e->errorMessage(message);
+                        addResult(lfn, group, true, message);
+                        e->Release();
+                        fileCanBeDeleted = false;
+                        break;
+                    }
+                    catch(...)
+                    {
+                        VStringBuffer message("Could not delete file %s from superfile %s: unknown exception", lfn, superOwner.queryLogicalName());
+                        addResult(lfn, group, true, message);
+                        fileCanBeDeleted = false;
+                        break;
+                    }
 
-    StringBuffer returnStr;
-    StringBuffer auditStr(",FileAccess,WsDfu,DELETED,");
-    IArrayOf<IEspDFUActionInfo> actionResults;
+                    if (removeEmptyOwningSuperFiles && (superOwner.numSubFiles(false)==0))
```

Comment:
I think this should also check fileCanBeDeleted, i.e. no point in continuing if fileCanBeDeleted failed.

Reply:
If the fileCanBeDeleted is set, the break will be called. Will this line not be reached ?

Reply:
Yes, my mistake.